engine/execution/ingestion/engine.go

```diff
@@ -139,214 +131,188 @@ func New(
+	blockID := block.ID()
+	log := e.log.With().Hex("block_id", blockID[:]).Logger()
+
+	_, err := e.execState.StateCommitmentByBlockID(ctx, blockID)
+	if err == nil {
+		// a statecommitment being stored indicates the block
+		// has been executed
+		log.Debug().Msg("block has been executed already")
+		return nil
+	}
 
-	block := &flow.Block{
-		Header:  proposal.Header,
-		Payload: proposal.Payload,
+	if !errors.Is(err, storage.ErrNotFound) {
+		return fmt.Errorf("could not query state commitment for block: %w", err)
 	}
 
-	e.metrics.StartBlockReceivedToExecuted(block.ID())
+	// unexecuted block
+	e.metrics.StartBlockReceivedToExecuted(blockID)
 
 	executableBlock := &entity.ExecutableBlock{
 		Block:               block,
 		CompleteCollections: make(map[flow.Identifier]*entity.CompleteCollection),
 	}
 
+	// acquiring the lock so that there is only one process modifying the queue
 	return e.mempool.Run(
 		func(
 			blockByCollection *stdmap.BlockByCollectionBackdata,
 			executionQueues *stdmap.QueuesBackdata,
-			orphanQueues *stdmap.QueuesBackdata,
 		) error {
-			// synchronize DB writing to avoid tx conflicts with multiple blocks arriving fast
-			err := e.blocks.Store(block)
-			if err != nil {
-				if !e.syncInProgress.Load() || !errors.Is(err, storage.ErrAlreadyExists) {
-					return fmt.Errorf("could not store block: %w", err)
-				}
-			}
+			// adding the block to the queue,
+			queue, added := enqueue(executableBlock, executionQueues)
 
-			// if block fits into execution queue, that's it
-			if queue, added := tryEnqueue(executableBlock, executionQueues); added {
-				e.log.Debug().Hex("block_id", logging.Entity(executableBlock)).Msg("added block to existing execution queue")
-				e.tryRequeueOrphans(executableBlock, queue, orphanQueues)
+			// if it's not added, it means the block is not a new block, it already
+			// exists in the queue, then bail
+			if !added {
+				log.Debug().Msg("block already exists in the execution queue")
 				return nil
 			}
 
-			// if block fits into orphan queues
-			if queue, added := tryEnqueue(executableBlock, orphanQueues); added {
-				e.log.Debug().Hex("block_id", logging.Entity(executableBlock)).Msg("added block to existing orphan queue")
-				e.tryRequeueOrphans(executableBlock, queue, orphanQueues)
-				// this is only queue which grew and could trigger threshold
-				if queue.Height() < e.syncModeThreshold {
-					return nil
-				}
-				if e.syncInProgress.CAS(false, true) {
-					// Start sync mode - initializing would require DB operation and will stop processing blocks here
-					// which is exactly what we want
-					e.StartSync(ctx, queue.Head.Item.(*entity.ExecutableBlock))
-				}
-				return nil
-			}
+			// whenever the queue grows, we need to check whether the state sync should be
+			// triggered.
+			firstUnexecutedHeight := queue.Head.Item.Height()
+			e.unit.Launch(func() {
+				e.checkStateSyncStart(firstUnexecutedHeight)
```

Comment:
Do we handle unit quit channel in this method?

Reply:
Good question, the `checkStateSyncStart` is blocking at the moment, so `e.unit.Launch` is already handling the quit channel